gandalf.js

```diff
@@ -1,47 +1,51 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?